metricbeat/module/prometheus/collector/collector_test.go

```diff
@@ -380,3 +384,140 @@ func TestSkipMetricFamily(t *testing.T) {
+	metricsPath := "/metrics"
+	server := initServer(metricsPath)
+	defer server.Close()
+
+	host := strings.TrimPrefix(server.URL, "http://")
+
+	config := map[string]interface{}{
+		"module":        "prometheus",
+		"metricsets":    []string{"collector"},
+		"hosts":         []string{server.URL},
+		"metrics_path":  metricsPath,
+		"metrics_count": true,
+	}
+
+	expectedEvents := 11
+
+	testCases := []struct {
+		name                string
+		expectedLabel       mapstr.M
+		expectedMetricCount int64
+	}{
+		{"Prod API Inf", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "le": "+Inf", "service": "api"}, 1},
+		{"Prod API 0.5", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "le": "0.5", "service": "api"}, 1},
+		{"Prod API 1", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "le": "1", "service": "api"}, 1},
+		{"Prod API Quantile 0.5", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "quantile": "0.5", "service": "api"}, 1},
+		{"Prod API Quantile 0.9", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "quantile": "0.9", "service": "api"}, 1},
+		{"Prod API Quantile 0.99", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "quantile": "0.99", "service": "api"}, 1},
+		{"Prod API", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "service": "api"}, 6},
+		{"Prod DB", mapstr.M{"environment": "prod", "instance": host, "job": "prometheus", "service": "db"}, 2},
+		{"Staging API", mapstr.M{"environment": "staging", "instance": host, "job": "prometheus", "service": "api"}, 2},
+		{"Staging DB", mapstr.M{"environment": "staging", "instance": host, "job": "prometheus", "service": "db"}, 2},
+		{"Default", mapstr.M{"instance": host, "job": "prometheus"}, 1},
+	}
+
+	f := mbtest.NewReportingMetricSetV2Error(t, config)
+	events, errs := mbtest.ReportingFetchV2Error(f)
+
+	for _, err := range errs {
+		t.Errorf("Unexpected error: %v", err)
+	}
+
+	assert.Equal(t, expectedEvents, len(events), "Number of events does not match expected")
+
+	sortPromEvents(events)
+
+	for i := range expectedEvents {
+		t.Run(testCases[i].name, func(t *testing.T) {
+			validateEvent(t, events[i], testCases[i].expectedLabel, testCases[i].expectedMetricCount)
+		})
+	}
+}
+func validateEvent(t *testing.T, event mb.Event, expectedLabels mapstr.M, expectedMetricsCount int64) {
+	t.Helper()
+
+	metricsCount, err := event.RootFields.GetValue("metrics_count")
+	assert.NoError(t, err, "Failed to get metrics_count")
+
+	labels, ok := event.RootFields["prometheus"].(mapstr.M)["labels"].(mapstr.M)
+	assert.True(t, ok, "Failed to get labels")
+
+	assert.Equal(t, expectedLabels, labels, "Labels do not match expected")
+	assert.Equal(t, expectedMetricsCount, metricsCount, "Metrics count does not match expected")
+}
+
+// NOTE(shmsr): If "test_histogram_bucket{environment="prod",service="api",le="0.1"} 0"
```

Comment:
Spent debugging this issue for hours on why tests are passing on the local machine (MacOS M1) but not on CI. I've written the details here. Please read it carefully; maybe the parser needs to be fixed. There's already a related issue opened for this one which I have linked.

cc: @gizas @devamanv

Reply:
Yep, exactly! I've already linked this in the comment itself of this particular test: https://github.com/elastic/beats/blob/42df5573e18da20d07f43123fd7d6a0ca00c2254/metricbeat/module/prometheus/collector/collector_test.go#L476

We have to fix the NaN handling.

Reply:
Oh, thanks, I missed that. Glad the issue is well known and we are tracking it already.